packages/eslint-plugin-turbo/lib/utils/calculate-inputs.ts

```diff
@@ -0,0 +1,449 @@
+            );
+          } else {
+            globalTasks[scriptName] = processTask(
+              rootTurboJson.workspacePath,
+              taskDefinition
+            );
+          }
+        }
+      );
+    }
+
+    this.projectWorkspaces.forEach((projectWorkspace) => {
+      if (!projectWorkspace.turboConfig) {
+        return;
+      }
+
+      Object.entries(projectWorkspace.turboConfig.pipeline).forEach(
+        ([taskName, taskDefinition]) => {
+          const { workspaceName: erroneousWorkspaceName, scriptName } =
+            getTaskAddress(taskName);
+          if (erroneousWorkspaceName) {
+            throw new Error(
+              "May not specify workspace name in non-root turbo.json"
+            );
+          }
+
+          const workspaceName = projectWorkspace.workspaceName;
+          workspaceTasks[workspaceName] = workspaceTasks[workspaceName] || {};
+          workspaceTasks[workspaceName][scriptName] = processTask(
+            projectWorkspace.workspacePath,
+            taskDefinition
+          );
+        }
+      );
+    });
+
+    return {
+      global,
+      globalTasks,
+      workspaceTasks,
+    };
+  }
+
+  getWorkspacePath(workspaceName: string): string | undefined {
+    return this.projectWorkspaces.find(
+      (workspaceConfig) => workspaceConfig.workspaceName === workspaceName
+    )?.workspacePath;
+  }
+
+  generateTestConfig(): ProjectTests {
+    return {
+      global: generateEnvironmentTest(
+        this._key.global,
+        this.projectRoot?.workspacePath
+      ),
+      globalTasks: Object.fromEntries(
+        Object.entries(this._key.globalTasks).map(([script, config]) => {
+          return [
+            script,
+            generateEnvironmentTest(config, this.projectRoot?.workspacePath),
+          ];
+        })
+      ),
+      workspaceTasks: Object.fromEntries(
+        Object.entries(this._key.workspaceTasks).map(
+          ([workspace, taskConfigs]) => {
+            const workspacePath = this.getWorkspacePath(workspace);
+            return [
+              workspace,
+              Object.fromEntries(
+                Object.entries(taskConfigs).map(([script, config]) => {
+                  return [
+                    script,
+                    generateEnvironmentTest(config, workspacePath),
+                  ];
+                })
+              ),
+            ];
+          }
+        )
+      ),
+    };
+  }
+
+  key() {
+    return this._key;
+  }
+
+  test(workspaceName: string | undefined, envVar: string) {
+    const tests = [
+      environmentTestArray(this._test.global),
+      ...Object.values(this._test.globalTasks).map((context) =>
+        environmentTestArray(context)
+      ),
+    ];
+
+    if (workspaceName) {
+      tests.push(
+        ...Object.values(this._test.workspaceTasks[workspaceName]).map(
+          (context) => environmentTestArray(context)
+        )
+      );
+    }
+
+    return tests.flat().some((test) => test(envVar));
+  }
```

Comment:
And this is the actual "test this key, please!" function. Most of the expensive things have been moved out of it; it just builds the list of tests needed to run, and then runs them.